cpp/src/transpose/transpose.cu

```diff
@@ -0,0 +1,176 @@
+ * limitations under the License.
+ */
+#include <cudf/transpose.hpp>
+#include <cudf/detail/transpose.hpp>
+#include <cudf/table/table_device_view.cuh>
+#include <cudf/utilities/nvtx_utils.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+#include <cudf/utilities/traits.hpp>
+#include <cudf/detail/copy.hpp>
+#include <utilities/integer_utils.hpp>
+
+namespace cudf {
+
+namespace {
+
+constexpr int WARP_SIZE = 32;
+constexpr int MAX_GRID_SIZE = (1<<16)-1;
+
+/**
+ * @brief Transposes the values from ncols x nrows input to nrows x ncols output 
+ * 
+ * @tparam T  Datatype of values pointed to by the views
+ * @param input[in]  Device view of input columns' data
+ * @param output[out]  Mutable device view of pre-allocated output columns' data
+ */
+template <typename T>
+__global__
+void gpu_transpose(table_device_view const input, mutable_table_device_view output)
+{
+  size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+
+  size_type stride_x = blockDim.x * gridDim.x;
+  size_type stride_y = blockDim.y * gridDim.y;
+
+  for (size_type i = x; i < input.num_columns(); i += stride_x) {
+    for (size_type j = y; j < input.num_rows(); j += stride_y) {
+      output.column(j).element<T>(i) = input.column(i).element<T>(j);
+    }
+  }
+}
+
+/**
+ * @brief Transposes the null mask from ncols x nrows input to nrows x ncols output
+ * 
+ * @tparam T  Datatype of values pointed to by the views
+ * @param input[in]  Device view of input columns' data
+ * @param output[out]  Mutable device view of pre-allocated output columns' data
+ */
+__global__
+void gpu_transpose_valids(table_device_view const input, mutable_table_device_view output)
+{
+  size_type x = blockIdx.x * blockDim.x + threadIdx.x;
+  size_type y = blockIdx.y * blockDim.y + threadIdx.y;
+
+  size_type stride_x = blockDim.x * gridDim.x;
+  size_type stride_y = blockDim.y * gridDim.y;
+
+  size_type i = x;
+  auto active_threads = __ballot_sync(0xffffffff, i < input.num_columns());
+  while (i < input.num_columns()) {
+    for (size_type j = y; j < input.num_rows(); j += stride_y) {
+      auto const result = __ballot_sync(active_threads, input.column(i).is_valid(j));
+
+      // Only one thread writes output
+      if (0 == threadIdx.x % WARP_SIZE) {
+        output.column(j).set_mask_word(word_index(i), result);
+      }
+    }
+
+    // Update active threads before branching
+    i += stride_x;
+    active_threads = __ballot_sync(active_threads, i < input.num_columns());
+  }
+}
+
+struct launch_kernel {
+  template <typename T>
+  void operator()(
+    table_view const& input,
+    mutable_table_view& output,
+    bool has_null,
+    cudaStream_t stream)
+  {
+    // Copy input columns `data` and `valid` pointers to device
+    auto device_input = table_device_view::create(input, stream);
+    auto device_output = mutable_table_device_view::create(output, stream);
+
+    dim3 dimBlock(WARP_SIZE, WARP_SIZE);
```

Comment:
TODO (not necessarily in this PR): benchmark, because a smaller block size (e.g. 32x8) may perform better. It would also require transposing via shared memory, which may improve performance anyway because it would enable coalesced loads *and* stores rather than one or the other.

Reply:
That makes sense. I added your comment to this PR for now.